buildSrc/src/main/java/org/elasticsearch/gradle/util/GradleUtils.java

```diff
@@ -120,4 +130,69 @@ public static void maybeConfigure(TaskContainer tasks, String name, Action<? sup
 
         return (Provider<T>) registration.getService();
     }
+
+    /**
+     * Add a source set and task of the same name that runs tests.
+     *
+     * IDEs are also configured if setup, and the test task is added to check. The new test source
+     * set extends from the normal test source set to allow sharing of utilities.
+     */
+    public static void addTestSourceSet(Project project, String sourceSetName) {
+        project.getPluginManager().apply(ElasticsearchJavaPlugin.class);
+
+        // create our test source set and task
+        SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+        SourceSet extraTestSourceSet = sourceSets.create(sourceSetName);
+        TaskProvider<Test> testTask = project.getTasks().register(sourceSetName, Test.class);
+        testTask.configure(task -> {
+            task.setTestClassesDirs(extraTestSourceSet.getOutput().getClassesDirs());
+            task.setClasspath(extraTestSourceSet.getRuntimeClasspath());
+        });
+        SourceSet mainSourceSet = sourceSets.getByName("main");
+        SourceSet testSourceSet = sourceSets.getByName("test");
+
+        extendConfiguration(project, testSourceSet, extraTestSourceSet, SourceSet::getCompileConfigurationName);
+        extendConfiguration(project, testSourceSet, extraTestSourceSet, SourceSet::getImplementationConfigurationName);
+        extendConfiguration(project, testSourceSet, extraTestSourceSet, SourceSet::getRuntimeConfigurationName);
+        extendConfiguration(project, testSourceSet, extraTestSourceSet, SourceSet::getRuntimeOnlyConfigurationName);
+
+        // tie this new test source set to the main and test source sets
+        Configuration extraTestCompileConfig = project.getConfigurations()
+            .getByName(extraTestSourceSet.getCompileClasspathConfigurationName());
+        Configuration extraTestRuntimeConfig = project.getConfigurations()
+            .getByName(extraTestSourceSet.getRuntimeClasspathConfigurationName());
+        extraTestSourceSet.setCompileClasspath(
+            project.getObjects().fileCollection().from(mainSourceSet.getOutput(), testSourceSet.getOutput(), extraTestCompileConfig)
```

Comment:
This could also just be `mainSourceSet.getOutput().plus(testSourceSet.getOutput()).plus(extraTestCompileConfig)`